ipu/examples/jax_ipu_mnist_neural_net_training.ipynb

```diff
@@ -0,0 +1,400 @@
+    "@partial(jax.jit, backend=\"ipu\", donate_argnums=(0,))\n",
+    "def update(params, x, y):\n",
+    "  grads = grad(loss)(params, x, y)\n",
+    "  return [(w - step_size * dw, b - step_size * db)\n",
+    "          for (w, b), (dw, db) in zip(params, grads)]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "execution": {
+     "iopub.execute_input": "2023-02-24T13:05:12.727232Z",
+     "iopub.status.busy": "2023-02-24T13:05:12.726987Z",
+     "iopub.status.idle": "2023-02-24T13:05:12.734676Z",
+     "shell.execute_reply": "2023-02-24T13:05:12.734094Z",
+     "shell.execute_reply.started": "2023-02-24T13:05:12.727212Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from torch.utils import data\n",
+    "from torchvision.datasets import MNIST\n",
+    "\n",
+    "def numpy_collate(batch):\n",
+    "  if isinstance(batch[0], np.ndarray):\n",
+    "    return np.stack(batch)\n",
+    "  elif isinstance(batch[0], (tuple,list)):\n",
+    "    transposed = zip(*batch)\n",
+    "    return [numpy_collate(samples) for samples in transposed]\n",
+    "  else:\n",
+    "    return np.array(batch)\n",
+    "\n",
+    "class NumpyLoader(data.DataLoader):\n",
+    "  def __init__(self, dataset, batch_size=1,\n",
+    "                shuffle=False, sampler=None,\n",
+    "                batch_sampler=None, num_workers=0,\n",
+    "                pin_memory=False, drop_last=False,\n",
+    "                timeout=0, worker_init_fn=None):\n",
+    "    super(self.__class__, self).__init__(dataset,\n",
+    "        batch_size=batch_size,\n",
+    "        shuffle=shuffle,\n",
+    "        sampler=sampler,\n",
+    "        batch_sampler=batch_sampler,\n",
+    "        num_workers=num_workers,\n",
+    "        collate_fn=numpy_collate,\n",
+    "        pin_memory=pin_memory,\n",
+    "        drop_last=drop_last,\n",
+    "        timeout=timeout,\n",
+    "        worker_init_fn=worker_init_fn)\n",
+    "\n",
+    "class FlattenAndCast(object):\n",
+    "  def __call__(self, pic):\n",
+    "    return np.ravel(np.array(pic, dtype=jnp.float32))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "execution": {
+     "iopub.execute_input": "2023-02-24T13:05:13.279641Z",
+     "iopub.status.busy": "2023-02-24T13:05:13.279433Z",
+     "iopub.status.idle": "2023-02-24T13:05:13.328832Z",
+     "shell.execute_reply": "2023-02-24T13:05:13.328033Z",
+     "shell.execute_reply.started": "2023-02-24T13:05:13.279623Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Define our dataset, using torch datasets\n",
+    "mnist_dataset = MNIST('/tmp/mnist/', download=True, transform=FlattenAndCast())\n",
+    "# IPU: dropping last micro batch with different shape.\n",
+    "training_generator = NumpyLoader(mnist_dataset, batch_size=batch_size, num_workers=0, drop_last=True)"
```

Comment:
note to us: remove this when doing "fast as possible" version

Reply:
i.e. use batch size that divides 50K

Reply:
We can in the infeed/outfeed version. For a general notebook like that, I kind of prefer to keep to show all the pitfalls and sharp bits!